rust/relay/src/server.rs

```diff
@@ -56,55 +169,215 @@ impl Server {
     fn handle_allocate_request(
-        &self,
+        &mut self,
         message: Message<Attribute>,
         sender: SocketAddr,
-    ) -> Option<(SocketAddr, Message<Attribute>)> {
-        tracing::debug!("Received TURN allocate request from: {sender}");
+    ) -> Result<(), ErrorCode> {
+        if self.allocations.contains_key(&sender) {
+            return Err(AllocationMismatch.into());
+        }
+
+        if self.used_ports.len() == MAX_AVAILABLE_PORTS as usize {
+            return Err(InsufficientCapacity.into());
+        }
+
+        let requested_transport = message
+            .get_attribute::<RequestedTransport>()
+            .ok_or(BadRequest)?;
+
+        if requested_transport.protocol() != UDP_TRANSPORT {
+            return Err(BadRequest.into());
+        }
+
+        let requested_lifetime = message.get_attribute::<Lifetime>();
 
-        let Some(_mi) = message.get_attribute::<MessageIntegrity>() else {
-            tracing::debug!("Turning down allocate request from {sender} because it is not authenticated");
+        let effective_lifetime = compute_effective_lifetime(requested_lifetime);
 
-            let mut message = Message::new(
-                MessageClass::ErrorResponse,
-                ALLOCATE,
-                message.transaction_id(),
-            );
-            message.add_attribute(ErrorCode::from(Unauthorized).into());
+        // TODO: Do we need to handle DONT-FRAGMENT?
+        // TODO: Do we need to handle EVEN/ODD-PORT?
 
-            return Some((sender, message));
+        let allocation = self.create_new_allocation();
+
+        let mut message = Message::new(
+            MessageClass::SuccessResponse,
+            ALLOCATE,
+            message.transaction_id(),
+        );
+
+        message.add_attribute(XorRelayAddress::new(allocation.ip4_relay_address.into()).into());
+        message.add_attribute(XorRelayAddress::new(allocation.ip6_relay_address.into()).into());
+        message.add_attribute(XorMappedAddress::new(sender).into());
+        message.add_attribute(effective_lifetime.clone().into());
+
+        self.pending_commands.push_back(Command::AllocateAddresses {
+            id: allocation.id,
+            ip4: allocation.ip4_relay_address,
+            ip6: allocation.ip6_relay_address,
+            expiry_in: effective_lifetime.lifetime(),
+        });
+        self.send_message(message, sender);
+
+        tracing::info!(
+            "Created new allocation for {sender} with address {} and lifetime {}s",
+            allocation.ip4_relay_address,
+            effective_lifetime.lifetime().as_secs()
+        );
+        tracing::info!(
+            "Created new allocation for {sender} with address {} and lifetime {}s",
+            allocation.ip6_relay_address,
+            effective_lifetime.lifetime().as_secs()
+        );
+
+        self.allocations.insert(sender, allocation);
+
+        Ok(())
```

Comment:
I think this function would be a bit easier to read if we separate it into 3 different helpers(signatures/naming might be a bit off):
`check_allocation_ok(&self, message: &Message<Attribute>)`
`create_new_allocation(&mut self) -> Allocation` (<- this one already exists)
`add_allocation_commands(&mut self, message: &Message<Attribute>, sender: &SocketAddr)`

Reply:
I still expect a lot of churn on the implementation of these functions but I'll keep it in mind to reduce them down so something more readable!